File: 290-word-pattern/word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] s_array = s.split(" ");

        if (s_array.length != pattern.length()) {
            return false;
        }

        Map<Character, String> char_map = new HashMap<>();
        Map<String, Character> word_map = new HashMap<>();
        
        for (int i=0; i < s_array.length; i++) {
            String word = s_array[i];
            Character c = pattern.charAt(i);

            if (word_map.containsKey(word) && word_map.get(word) != c) {
                return false;
            }

            if (char_map.containsKey(c) && !char_map.get(c).equals(word)) {
                return false;
            }

            word_map.put(word, c);
            char_map.put(c, word);
        }

        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        s_array = s.split(" ")
        if len(s_array) != len(pattern):
            return False

        s_set = set()

        pattern_map = {}
        for i in range(len(pattern)):
            c = pattern[i]
            if c in pattern_map:
                if pattern_map[c] != s_array[i]:
                    return False
            else:
                if s_array[i] in s_set:
                    return False

            pattern_map[c] = s_array[i]
            s_set.add(s_array[i])
            
        return True

